core/jvm/src/test/scala/fs2/async/SignalTestLaws.scala

```diff
@@ -0,0 +1,142 @@
+      val finiteStream = Stream.emits(finiteElems).covary[IO]
+      unsafeHold(firstElem, finiteStream)
+    }
+    Arbitrary(gen.map(_.unsafeRunSync()))
+  }
+
+  private type SignalIO[A] = immutable.Signal[IO, A]
+
+  /**
+    * We need an instance of Eq for the Discipline laws to work, but actually
+    * running a Signal is effectful, so we have to resort to unsafely
+    * performing the effect inside the equality check to have a valid Eq
+    * instance.
+    *
+    * Moreover, equality of Signals is kind of murky. Since the exact discrete
+    * stream and gets that you see are non-deterministic even if two observers
+    * are looking at the same Signal, we need some notion of equality that is
+    * robust to this non-determinism.
+    *
+    * We say that two Signals are equal if they converge to the same value. And
+    * two streams converge to the same value if the "last" element of their
+    * discrete streams match (where "last" means either truly the last element
+    * if the discrete stream is finite or the last element seen if we pause
+    * changes to the Signal for a while), calling get after the last element
+    * of the discrete stream results in a match with the last element, and the
+    * first (or any) element of the continuous stream called after the "last"
+    * element of the discrete stream also matches.
+    *
+    * We will be testing with Signals generated by [[unsafeSignalArbitrary]],
+    * which are always finite Signals, so we can just take the actual last
+    * element of the discrete stream instead of doing some time-based wait for
+    * the Signal to settle.
+    */
+  private implicit def unsafeSignalEquality[A: Eq]: Eq[SignalIO[A]] = new Eq[SignalIO[A]] {
+    override def eqv(x: SignalIO[A], y: SignalIO[A]): Boolean = {
+      val action = for {
+        lastDiscreteX <- x.discrete.compile.last.map(_.get)
+        lastDiscreteY <- y.discrete.compile.last.map(_.get)
+        retrievedX <- x.get
+        retrievedY <- y.get
+        aContinuousX <- x.continuous.head.compile.last.map(_.get)
+        aContinuousY <- y.continuous.head.compile.last.map(_.get)
+      } yield {
+        val lastDiscretesAreSame = Eq[A].eqv(lastDiscreteX, lastDiscreteY)
+        val lastGetsAreTheSame = Eq[A].eqv(retrievedX, retrievedY)
+        val continuousAfterGetIsTheSame = Eq[A].eqv(aContinuousX, aContinuousY)
+        val lastDiscreteAgreesWithGet = Eq[A].eqv(lastDiscreteX, retrievedX)
+        val continuousAfterGetAgreesWithGet = Eq[A].eqv(aContinuousX, retrievedX)
+
+        lastDiscretesAreSame &&
+        lastGetsAreTheSame &&
+        continuousAfterGetIsTheSame &&
+        lastDiscreteAgreesWithGet &&
+        continuousAfterGetAgreesWithGet
+      }
+      action.unsafeRunSync()
+    }
+  }
+
+  checkAll(
+    "immutable.Signal",
+    FunctorTests.apply[SignalIO](immutable.Signal.signalIsFunctor).functor[String, Int, Double]
+  )
+
+  private implicit val testExecutionContext: ExecutionContextExecutor =
```

Comment:
Can we either use the global EC or shut this one down at the end of all tests? Maybe TestUtil?

Reply:
Hmmm... I'd like to keep as many factors constant when testing as possible and the `global` will spin up threads based on the number of processors. If you don't think that's too big of a deal I'm happy to use the global one, but for now I'll just shutdown the executor. If you have preferences around what you'd like to move to `TestUtil` let me know as well.

Reply:
Never mind, see my next comment.